timemachine/cpp/src/gpu_utils.cuh

```diff
@@ -124,37 +125,39 @@ __device__ __forceinline__ void energy_warp_reduce(volatile __int128 *shared_dat
 // values repeatedly until the final warp level reduction which stores the final accumulated value in `shared_data[0]`.
 // For more details reference the PDF in the docstring for k_accumulate_energy
 template <unsigned int THREADS_PER_BLOCK>
-__device__ __forceinline__ void block_energy_reduce(volatile __int128 *shared_data, int tid) {
+__device__ __forceinline__ void block_energy_reduce(
+    volatile __int128 *shared_data, // [THREADS_PER_BLOCK]
+    int tid) {
     static_assert(THREADS_PER_BLOCK <= 1024 && (THREADS_PER_BLOCK & (THREADS_PER_BLOCK - 1)) == 0);
 
     // For larger blocks larger than a warp
     if (THREADS_PER_BLOCK >= 1024) {
-        if (tid < 512) {
+        if (tid < 512 && tid + 512 < THREADS_PER_BLOCK) {
```

Comment:
Can probably remove some of these, will revisit to make this more strict. Think this only matters where thread block is <32 and a power of two